### PR TITLE
EWASM interface tests

### PR DESCRIPTION
**Pull request by lrettig**
WIP. All finished and working except `log` test. Still need to test with metering on.

CF https://github.com/ethereum/tests/pull/394.

### CODE REVIEW

**Comment by lrettig** [comment]
@axic Confirmed that all tests work with metering on (via https://github.com/ethereum/tests/pull/397) except one test (`getExternalCodeSize`) which has a hardcoded code size parameter.

**Comment by lrettig** [comment]
Squashed all my commits into a single clean commit, back to you @jwasinger

File: src/GeneralStateTestsFiller/stEWASMTests/utils/callCodeFiller.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          (memory 1)
+          (data (i32.const 0)  "\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\ef\be\ad\de")
+
+          (export "memory" (memory 0))
+          (export "main" (func $main))
+
+          (func $main
+            (i32.store (i32.const 96) (call $callCode (i64.const 100000) (i32.const 0) (i32.const 32) (i32.const 32) (i32.const 1) (i32.const 64) (i32.const 32)))
+
+            ;; Expect that the call did not trap i.e. result was 1
```
Comment (jwasinger) [inline src/GeneralStateTestsFiller/stEWASMTests/utils/callCodeFiller.yml]: Trapping here does nothing in terms of affecting the state and changing the test outcome.  Maybe add a store statement after this and add an expect case for `abcdabcdabcdabcdabcdabcdabcdabcdabcdabcd` to have storage after trap statement be set.

File: src/GeneralStateTestsFiller/stEWASMTests/utils/callDataCopyFiller.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      nonce: ''
+      storage: {}
+    # main contract, tx receiver
+    abcdabcdabcdabcdabcdabcdabcdabcdabcdabcd:
+      balance: '100000000000'
+      code: |
+        (module
+          (import "ethereum" "callDataCopy" (func $callDataCopy (param i32 i32 i32)))
+          (import "ethereum" "storageStore" (func $storageStore (param i32 i32)))
+          (import "debug" "printMemHex" (func $printMemHex (param i32 i32)))
```
Comment (jwasinger) [inline src/GeneralStateTestsFiller/stEWASMTests/utils/callDataCopyFiller.yml]: IMO let's try and move debug statements out of tests after we have ensured that the WAST is written properly. The code is easier to read and understand when there aren't debug statements in it.

File: src/GeneralStateTestsFiller/stEWASMTests/utils/callCodeFiller.yml (hunk 1 above)
Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/utils/callCodeFiller.yml]: "trap" is the wrong word here. Basically I think we should check that `callCode` had a result of 1 to indicate success. I think your comment here refers to the conditional that follows and the `(unreachable)`. I got this syntax from the `ewasm-kernel` tests e.g. https://github.com/ewasm/ewasm-kernel/blob/master/tests/interface/call.wast. Is there a reason those tests include this syntax but it's not useful here?

File: src/GeneralStateTestsFiller/stEWASMTests/utils/callDataCopyFiller.yml (hunk 2 above)
Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/utils/callDataCopyFiller.yml]: Agreed, good catch, will fix

**Comment by lrettig** [comment]
@jwasinger I fixed these issues and rebased. No more `unreachable` (I left one since the following code relies on an invariant, like an assert), instead everything is being tested in the state tests now. Also cleaned up some leftover debugging and added some locals to make the complex tests more readable. Back to you.

File: src/GeneralStateTestsFiller/stEWASMTests/utils/revertFiller.yml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        (module
+          (import "ethereum" "revert" (func $revert (param i32 i32)))
+          (memory 1)
+          (export "memory" (memory 0))
+          (export "main" (func $main))
+          (func $main
+            ;; Store some data to return
+            (i32.store (i32.const 0) (i32.const 0xabcd))
+            ;; Revert and return it
+            (call $revert
```
Comment (jwasinger) [inline src/GeneralStateTestsFiller/stEWASMTests/utils/revertFiller.yml]: Could you add a state changing operation before the `revert` statement?  And the expect should check that the state change was reverted.

Comment (jwasinger) [reply src/GeneralStateTestsFiller/stEWASMTests/utils/revertFiller.yml]: Also add some state changing operation afterwards, and then add another expect condition that ensures that the second state changing operation was not executed.

Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/utils/revertFiller.yml]: Done

**Comment by lrettig** [comment]
@jwasinger I fixed the test to expect the correct return value from the reverted `call`, and removed the gas checks as we discussed. Back to you again.

Let me know if you want me to squash.

**Pull request merged**